src/aws/flb_aws_util.c

```diff
@@ -927,3 +933,84 @@ flb_sds_t flb_get_s3_key(const char *format, time_t time, const char *tag,
         }
         return NULL;
 }
+
+/*
+ * Count the substring number in string.
+ */
+int count_substring(char *str1, char *str2){
+    int x = 0;
+    char *p;
+
+    do {
+        p = strstr(str1, str2);
+        if (p != NULL) {
+            str1 = p + 1;
+            x = x + 1;
+        }
+    }
+    while(p != NULL);
+    return x;
+}
+
+/*
+ * Adjust the time_key output format according to time_key_format.
+ */
+int create_time_format_string(char **out_buff, size_t *out_size,
+                              char *time_key_format, const struct tm *timeptr,
+                              size_t len, struct flb_time *tms)
+{
+    char millisecond_string[FLB_MILLISECOND_STRING_LENGTH+1];
+    char nanosecond_string[FLB_NANOSECOND_STRING_LENGTH+1];
+    char *tmp_buff_time_str;
+    char *mem_for_time_key_format;
+    char *tmp_out_buff;
+    size_t tmp_out_size;
+
+    /* guess the maximum length needed for out_buff, replace %F to 3 chars and %L to 9 chars */
+    len = 12*strlen(time_key_format)+count_substring(time_key_format, FLB_MILLISECOND_STRING)+7*count_substring(time_key_format, FLB_NANOSECOND_STRING);
+    mem_for_time_key_format = (char *)calloc(strlen(time_key_format), sizeof(char));
+    if (!mem_for_time_key_format) {
+        return -1;
+    }
+    tmp_buff_time_str = (char *)calloc(len, sizeof(char));
+    if (!tmp_buff_time_str) {
+        return -1;
+    }
+    *out_buff = (char *)calloc(len, sizeof(char));
+    if (!*out_buff) {
+        return -1;
+    }
+    tmp_out_buff = (char *)calloc(len, sizeof(char));
+    if (!tmp_out_buff) {
+        return -1;
+    }
+    memcpy(mem_for_time_key_format, time_key_format, strlen(time_key_format)+1);
+    snprintf(millisecond_string, FLB_MILLISECOND_STRING_LENGTH+1, "%" PRIu64, (uint64_t) tms->tm.tv_nsec / 1000000);
+    snprintf(nanosecond_string, FLB_NANOSECOND_STRING_LENGTH+1, "%" PRIu64, (uint64_t) tms->tm.tv_nsec);
+    /* replace %F to millisecond, %L to nanosecond in time_key_format */
+    for (int i = 0; i < strlen(time_key_format); i++) {
+        if (!strncmp(time_key_format+i, FLB_MILLISECOND_STRING, strlen(FLB_MILLISECOND_STRING))) {
+            strncat(tmp_buff_time_str, millisecond_string, strlen(millisecond_string));
+            i += strlen(FLB_MILLISECOND_STRING)-1;
+        } else if (!strncmp(time_key_format+i, FLB_NANOSECOND_STRING, strlen(FLB_NANOSECOND_STRING))) {
+            strncat(tmp_buff_time_str, nanosecond_string, strlen(nanosecond_string));
+            i += strlen(FLB_NANOSECOND_STRING)-1;
+        } else {
+            strncat(tmp_buff_time_str,time_key_format+ i,1);
+        }
+    }
+    memcpy(time_key_format, tmp_buff_time_str, strlen(tmp_buff_time_str)+1);
```

Comment:
why do you copy the processed result string to time_key_format?

Reply:
It's bad practice to modify the function parameters unless the user is expecting the output to be set to the param